doc/source/ray-core/examples/rdt/grpo_contextual_bandits.py

```diff
@@ -0,0 +1,461 @@
+        out = self.linear(out)
+        return residual + out
+
+
+class ResidualMLP(torch.nn.Module):  # Sized to ~50 MB of parameters.
+    """Model used for Generator and Learner.
+
+    It takes a 2D state vector as input and produces logits for each action.
+    """
+
+    def __init__(self, hidden_dim: int = 512, depth: int = 50):
+        super().__init__()
+        self.input = torch.nn.Linear(STATE_DIM, hidden_dim, bias=True)
+        self.backbone = torch.nn.ModuleList(
+            ResidualBlock(hidden_dim) for _ in range(depth - 1)
+        )
+        self.head = torch.nn.Linear(hidden_dim, ACTION_DIM, bias=True)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.input(x)
+        for block in self.backbone:
+            x = block(x)
+        x = self.head(x)
+        return x
+
+
+# -- Utilities --
+def sample_unit_vector(batch_size: int, dim: int = STATE_DIM) -> torch.Tensor:
+    """Sample unit vectors of shape [batch_size, dim] by normalizing Gaussian draws."""
+    assert batch_size > 1, "Batch size must be greater than 1"
+    v = torch.randn(batch_size, dim)
+    norms = v.norm(dim=-1, keepdim=True) + 1e-8
+    return v / norms
+
+
+# -- Actors --
+@ray.remote
+class ReplayBuffer:
+    """Storage for scored trajectory slices.
+
+    This class stores the past experiences (AKA trajectories, or slices) of the model.
+    This allows the learner to sample and learn from the same experiences multiple times
+    by comparing the latest model with previous models.
+
+    The sampler weights the trajectories by the policy version, such that trajectories produced
+    by more recent versions of the model are more likely to be sampled.
+    """
+
+    def __init__(self) -> None:
+        # Each entry stores a TrajectorySlice with CPU tensors.
+        self.storage: list[TrajectorySlice] = []
+
+    def put(self, slice: TrajectorySlice) -> None:
+        """Add a new slice to the buffer.
+
+        The buffer discards the oldest slices if the buffer gets too large to prevent memory leaks,
+        and so that the latest model can gradually explore further from the initial random policy.
+        """
+        self.storage.append(slice)
+        if len(self.storage) > MAX_BUFFER_SIZE:
+            self.storage = self.storage[-MAX_BUFFER_SIZE:]
+
+    def sample_from(self, n: int) -> list[TrajectorySlice]:
+        """Sample n scored trajectory slices."""
+        # The probability of sampling a slice is proportional to its policy version.
+        total = sum(slice["policy_version"] for slice in self.storage)
+        probs = [slice["policy_version"] / total for slice in self.storage]
+        # Sample with replacement without exceeding the buffer's size.
+        n = min(n, self.size())
+        chosen = np.random.choice(self.size(), size=n, p=probs, replace=True)
+        return [self.storage[i] for i in chosen]
+
+    def size(self) -> int:
+        return len(self.storage)
+
+
+@ray.remote(num_gpus=1)
```

Comment:
Actually can we update the script to not use GPU for scorer? Seems simpler this way.